src/ggml.c

```diff
@@ -16756,59 +16654,80 @@ int ggml_graph_compute(struct ggml_cgraph * cgraph, struct ggml_cplan * cplan) {
+        .cgraph                  = cgraph,
+        .cplan                   = cplan,
+        .perf_node_start_cycles  = 0,
+        .perf_node_start_time_us = 0,
+        .n_threads               = n_threads,
+        .node_n                  = -1,
+        .abort_callback          = NULL,
+        .abort_callback_data     = NULL,
     };
     struct ggml_compute_state * workers = alloca(sizeof(struct ggml_compute_state)*n_threads);
 
-    // create thread pool
-    if (n_threads > 1) {
-        for (int j = 1; j < n_threads; ++j) {
-            workers[j] = (struct ggml_compute_state) {
-                .thrd   = 0,
-                .ith = j,
-                .shared = &state_shared,
-            };
+    const int * n_tasks_arr = cplan->n_tasks;
 
-            const int rc = ggml_thread_create(&workers[j].thrd, NULL, ggml_graph_compute_thread, &workers[j]);
-            GGML_ASSERT(rc == 0);
-        }
+    // create thread pool arguments
+    for (int j = 1; j < n_threads; ++j) {
+        workers[j] = (struct ggml_compute_state) {
+            .thrd   = 0,
+            .ith = j,
+            .shared = &state_shared,
+        };
     }
     workers[0].ith = 0;
     workers[0].shared = &state_shared;
 
-    const int64_t perf_start_cycles  = ggml_perf_cycles();
-    const int64_t perf_start_time_us = ggml_perf_time_us();
+    int compute_status = GGML_EXIT_SUCCESS;
+    for (int node_n = 0; node_n < cgraph->n_nodes; ++ node_n) {
+        if (cplan->abort_callback && cplan->abort_callback(cplan->abort_callback_data)) {
+            compute_status = GGML_EXIT_ABORTED;
+            break;
+        }
 
-    // this is a work thread too
-    int compute_status = (size_t) ggml_graph_compute_thread(&workers[0]);
+        struct ggml_tensor* node = cgraph->nodes[node_n];
 
-    // don't leave affinity set on the main thread
-    clear_numa_thread_affinity();
+        state_shared.perf_node_start_cycles  = ggml_perf_cycles();
+        state_shared.perf_node_start_time_us = ggml_perf_time_us();
+
+        struct ggml_compute_params params = {
+            /*.type  =*/ -1,
+            /*.ith   =*/ 0,
+            /*.nth   =*/ n_tasks_arr[node_n],
+            /*.wsize =*/ cplan->work_size,
+            /*.wdata =*/ cplan->work_data,
+        };
 
-    // join or kill thread pool
-    if (n_threads > 1) {
-        for (int j = 1; j < n_threads; j++) {
-            const int rc = ggml_thread_join(workers[j].thrd, NULL);
-            GGML_ASSERT(rc == 0);
+        if (GGML_OP_HAS_INIT[node->op]) {
+            params.type = GGML_TASK_INIT;
+            ggml_compute_forward(&params, node);
         }
+
+        if (n_tasks_arr[node_n] > 1) {
+            state_shared.node_n = node_n;
+            for (int j = 1; j < n_tasks_arr[node_n]; ++j)
+                tpool_add_work(tpool, ggml_graph_compute_thread, &workers[j]);
+        }
+
+        params.type = GGML_TASK_COMPUTE;
+        ggml_compute_forward(&params, node);
```

Comment:
If I understand the code correctly the main thread managing the other threads still does part of the computation itself. A design where the main thread queues work for all worker threads and then waits for them to be done would maybe be better since this would allow you to parallelize the computations with administrative overhead (and maybe computations from multiple tensors, but I think the overhaul by @slaren also does something like that). But this can be done in another PR.

Reply:
The work scheduling should be intentionally identical to current master to ease comparison (modulo errors in my understanding of it).